contracts/core/EntryPoint.sol

```diff
@@ -680,30 +675,62 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuard,
                     }
                 }
             }
-            actualGas += preGas - gasleft();
+            uint postOpGas = preGas - gasleft();
+            _postOpGas(postOpGas);
+            actualGas += postOpGas;
             actualGasCost = actualGas * gasPrice;
             if (opInfo.prefund < actualGasCost) {
                 revert FailedOp(opIndex, "AA51 prefund below actualGasCost");
             }
             uint256 refund = opInfo.prefund - actualGasCost;
             _incrementDeposit(refundAddress, refund);
             bool success = mode == IPaymaster.PostOpMode.opSucceeded;
-            emit UserOperationEvent(
-                opInfo.userOpHash,
-                mUserOp.sender,
-                mUserOp.paymaster,
-                mUserOp.nonce,
+            _emitUserOperationEvent(
+                opInfo,
                 success,
                 actualGasCost,
                 actualGas
             );
         } // unchecked
     }
 
-    /**
-     * The gas price this UserOp agrees to pay.
-     * Relayer/block builder might submit the TX with higher priorityFee, but the user should not.
-     * @param mUserOp - The userOp to get the gas price from.
+    //solhint-disable-next-line no-empty-blocks
+    function _postOpGas(uint256 postOpGas) internal virtual {
```

Comment:
I am wondering, is it really the only time we may want to expose an internal variable to the simulation code? How come?

Reply:
We want to expose anything that is gas-related

Reply:
Maybe call it `simulationHookPostOpGas()` or something similar, as this is what it is? Plus, maybe we want more hooks like this one in other places for simulation?